### ISSUE TITLE
*Datatable* - Date fields do not take into account Locale Settings

### ISSUE THREAD

**Issue by florebloes** [alexed1/LightningFlowComponents#847]
### Steps to reproduce

1. Screen Flow - add datatable component, display in table a date field (installed version: 3.3)
2. When in a record, type in date column for a line a date eg. 16th october
3. Save 
4. You will see that the date being displayed in the datatable is the 15th of October, however, when you navigate to the actual record, it's save as the 16th of October

note - I am not using any special attributes in the datatable

### Expected behaviour
Datatable component should take into account locale settings for date & time. 
(checked the V2 code, here I saw you do use type "date locale", however, not sure if you are using this in the newest release as well ? ) 

### Actual behaviour
Displays date - 1 day in the component, while it is being saved in the background as the actual entered date when you navigate into the actual record . 

### Screenshots
Datatable view before save: 
![image](https://user-images.githubusercontent.com/88318233/128686641-c70cbe93-0bff-40ef-b75c-2a3f2e8ca370.png)

Datatable view after save:
![image](https://user-images.githubusercontent.com/88318233/128686808-d6d1f208-4e8d-4071-8c58-0e2b8092625c.png)

Record field in the record:
![image](https://user-images.githubusercontent.com/88318233/128687076-dd82c1d5-9d1c-4673-adb0-e9075f588a28.png)

**Comment by ericrsmith35** [comment]
The component uses a time-zone offset value for the running user to display dates correctly when the records are first read into the component.  Any field edits will show as the internal system values until the records are read back into a new datatable.

**Comment by florebloes** [comment]
@ericrsmith35 not clear to me. It's a screen flow I embedded on the record page. So when navigating back and forth to the record, the flow constantly gets triggered and each time gets the quote line items. So the edits are each time read back into the datatable as the flow gets 'loaded' again. 

Here is overview of entire flow : 
![image](https://user-images.githubusercontent.com/88318233/128716191-2a2d778a-3d27-4ee7-be1b-d45af38f5a6b.png)

**Comment by florebloes** [comment]
Hi @ericrsmith35 could you please review my question again!  Thanks in advance!

**Comment by ericrsmith35** [comment]
It looks like the correct date is being saved in the record.  Are you seeing the wrong date the next time the records are loaded and displayed in the datatable?

**Comment by florebloes** [comment]
@ericrsmith35 yes. The flow is just a component on my lightning record page. So each time I navigate back to the record, the flow is displayed, meaning each time the data is loaded back into the datable

**Comment by florebloes** [comment]
@ericrsmith35 Any feedback Yet?

**Comment by Dietermartens** [comment]
@ericrsmith35 I'm having the same issue here... our dates are correct on the record, but in the datatable they are displayed minus one day (it looks like it's converted to datetime and then displayed in UTC).

**Comment by ericrsmith35** [comment]
Can you share some screenshots of this happening and what timezone you are
in and what time of day this is happening?

**Comment by Dietermartens** [comment]
Hi Eric, 

We are base GMT+2 but it should not matter because it's a DATE field and not a DATATIME. It's actually as @florebloes described and showed, on the record, the date is saved and displayed correctly and in the datatable it is always -1 day. Our Salesforce is on userlevel and system eleven configured to use GMT+2. 

As far as I can see it's always happening

**Comment by florebloes** [comment]
@ericrsmith35 any further feedback on this?

**Comment by florebloes** [comment]
Hi Erik, here is the solution to the problem. Works now after this adjustment in the code

WHAT? Datatable for planning is showing wrong date (minus 1)
A screen flow using datatable was created, however in the javascript of this component there needs some code adjustment because else the date is shown minus -1 day in the datatable itself.

lwc component : database

file : database.js => comment/remove the part that is commented in the snippet below. This date conversion happening here is not necessary anymore. (was a bug in the standard sf datatable component)

`     // Adjust date with offset based on User's timezone
            dateFields.forEach(date => {
                if (record[date]) {
                   let dt = Date.parse(record[date]);
                    let d = new Date();
                    record[date] = new Date(d.setTime(Number(dt) - Number(this.timezoneOffset)));
                }
            }); `

**Comment by ericrsmith35** [comment]
I see the opposite issue in v3.2.2.  The table shows the correct date and
the record stores the correct date.  If I comment out that section, the
record still stores the correct date, but the table displays one day
earlier.

**Comment by ericrsmith35** [comment]
@florebloes @Dietermartens 
I have implemented a bug fix for the Datatable component where dates could sometimes be off by 1 day.  This would typically happen when the date being displayed, and the current date had a mismatch with one of them being in Daylight Savings time and the other one was not.

Before publicly publishing this fix, I would like to get feedback from users that had experienced the issue.  Please install the current versions of the UnofficialSF Base Packs (https://unofficialsf.com/flow-action-and-screen-component-basepacks/) then install Version 4.0.4 of the Datatable component.  

Sandbox (https://test.salesforce.com/packaging/installPackage.apexp?p0=04t5G000003rUxRQAU)
Production or Dev (https://login.salesforce.com/packaging/installPackage.apexp?p0=04t5G000003rUxRQAU)

Please let me know if this fixes your issue or if you are still having a problem with dates being off by one day.

Thanks,
Eric Smith

**Comment by frank-vdp** [comment]
Hi Eric,

Your fix has rectified my problem...Due Date before update:
[image: image.png]

...Due Date after update:
[image: image.png]

Thanks for this Datatable component - I use it everywhere!

Cheers,
Frank
